src/test/kotlin/com/terraformation/backend/funder/FundingEntityServiceTest.kt

```diff
@@ -0,0 +1,263 @@
+import com.terraformation.backend.customer.event.UserDeletionStartedEvent
+import com.terraformation.backend.customer.model.TerrawareUser
+import com.terraformation.backend.db.DatabaseTest
+import com.terraformation.backend.db.funder.FundingEntityId
+import com.terraformation.backend.db.funder.tables.pojos.FundingEntitiesRow
+import com.terraformation.backend.db.funder.tables.references.FUNDING_ENTITY_USERS
+import com.terraformation.backend.funder.db.FundingEntityExistsException
+import com.terraformation.backend.funder.db.FundingEntityNotFoundException
+import com.terraformation.backend.funder.db.FundingEntityStore
+import com.terraformation.backend.funder.db.FundingEntityUserStore
+import com.terraformation.backend.mockUser
+import io.mockk.every
+import java.time.Instant
+import org.junit.jupiter.api.Assertions.assertEquals
+import org.junit.jupiter.api.Assertions.assertNull
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.assertThrows
+import org.springframework.security.access.AccessDeniedException
+
+class FundingEntityServiceTest : DatabaseTest(), RunsAsUser {
+  override val user: TerrawareUser = mockUser()
+
+  private val clock = TestClock()
+  private lateinit var fundingEntityStore: FundingEntityStore
+  private lateinit var fundingEntityUserStore: FundingEntityUserStore
+  private lateinit var fundingEntityId: FundingEntityId
+  private lateinit var service: FundingEntityService
+
+  @BeforeEach
+  fun setUp() {
+    fundingEntityStore = FundingEntityStore(dslContext)
+    fundingEntityUserStore = FundingEntityUserStore(dslContext)
+    service = FundingEntityService(clock, dslContext, fundingEntityStore, fundingEntityUserStore)
+
+    fundingEntityId = insertFundingEntity()
+    insertOrganization()
+
+    every { user.canReadFundingEntities() } returns true
+    every { user.canCreateFundingEntities() } returns true
+    every { user.canDeleteFundingEntities() } returns true
+    every { user.canUpdateFundingEntities() } returns true
+    every { user.canUpdateFundingEntityProjects() } returns true
+  }
+
+  @Test
+  fun `create requires user to be able to manage funding entities`() {
+    every { user.canCreateFundingEntities() } returns false
+
+    assertThrows<AccessDeniedException> { service.create("Some Other Entity") }
+  }
+
+  @Test
+  fun `create populates created and modified fields`() {
+    val newTime = clock.instant().plusSeconds(1000)
+    clock.instant = newTime
+
+    val newUserId = insertUser()
+
+    val name = "New Funding Entity"
+    val row = FundingEntitiesRow(name = name)
+
+    every { user.userId } returns newUserId
+
+    val createdModel = service.create(row.name!!)
+
+    val expected =
+        row.copy(
+            createdBy = user.userId,
+            createdTime = clock.instant(),
+            id = createdModel.id,
+            modifiedBy = newUserId,
+            modifiedTime = clock.instant(),
+            name = name,
+        )
```

Comment:
Two things here.

I'm not sure what value `row` is adding here since it's just initialized with one field and that field is overridden in the `copy()` call.  Could get rid of `row` and do `service.create(name)` instead, then turn `row.copy()` into `FundingEntitiesRow()`.

Second, this could use `FundingEntitiesRecord` instead of `FundingEntitiesRow`, at which point you can use `assertTableEquals` instead of having to explicitly fetch the actual row.

The difference between the two is that the `Record` classes are a bit more heavyweight, but part of the extra weight is that they include additional metadata such as which table they came from. We're able to use that in `assertTableEquals` to automatically query the right table.

Reply:
It seems that to use `assertTableEquals`, the table has to be empty otherwise, so to use that I'd have to also get the record for the entity that is inserted in the `@BeforeEach` setup. Is that preferred? Seems like the same number of queries either way 🤔

Reply:
Sorry, I forgot I had everything I need to create the record for the initial entity; will get this updated